### PR TITLE
[WIP] Update VM based templates with private IP and update Tiltfile

### PR DESCRIPTION
**Pull request by nawazkh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- This PR updates the templates (based off of VMs) to use private IP of the API Server of the control plane of the workload cluster for inter-node communication.
- This PR also updates the Tiltfile with logic to peer AKS VNets with workload cluster's VNets.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/5261

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Update VM based templates to use Private IP solution for workload cluster's inter-node communication. Also update Tiltfile to update AKS Vnet with workload cluster's VNet
```

### CODE REVIEW

**Review by mboersma — COMMENTED** [review]
This seems like a reasonable approach. Should we have somewhere to document which CIDR ranges are "reserved" by which flavors? If we add a new flavor, we'd want to make sure it doesn't overlap, and it would be better not to have that rely just on maintainers' memories.

**Comment by nawazkh** [comment]
Lemme break this PR into multiple smaller PRs updating each of the templates. I will ship this PR with changes to default template and addressing the comments.

File: Tiltfile [hunk 1]
```diff
@@ -317,9 +319,14 @@ def flavors():
...
 
+    delete_all_workload_clusters = kubectl_cmd + " delete clusters --all --wait=false"
+
+    if "aks" in settings.get("kustomize_substitutions", {}).get("MGMT_CLUSTER_NAME", ""):
+        delete_all_workload_clusters += clear_aks_vnet_peerings()
+
     local_resource(
         name = "delete-all-workload-clusters",
-        cmd = kubectl_cmd + " delete clusters --all --wait=false",
+        cmd = ["sh", "-ec", delete_all_workload_clusters],
```
[the remark this answers is not in the document]
Comment (nawazkh) [reply Tiltfile:329]: I think we do need `sh` here. We are updating `delete_all_workload_clusters` to have a bunch of delete vnet peering commands added in this function https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/5248/files#diff-c2ee8653e1d6b85f0aadf87cd438a9250806c052877248442be4d434cbc52425R522-R532

File: Tiltfile [hunk 2]
```diff
@@ -382,17 +389,30 @@ def deploy_worker_templates(template, substitutions):
...
+    flavor_cmd += "; echo \"API Server of $CLUSTER_NAME is accessible\""
 
     # copy the kubeadm configmap to the calico-system namespace.
     # This is a workaround needed for the calico-node-windows daemonset to be able to run in the calico-system namespace.
     if "windows" in flavor_name:
         flavor_cmd += "; until " + kubectl_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig get configmap kubeadm-config --namespace=kube-system > /dev/null 2>&1; do sleep 5; done"
         flavor_cmd += "; " + kubectl_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig create namespace calico-system --dry-run=client -o yaml | " + kubectl_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig apply -f -; " + kubectl_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig get configmap kubeadm-config --namespace=kube-system -o yaml | sed 's/namespace: kube-system/namespace: calico-system/' | " + kubectl_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig apply -f -"
 
+    # TODO: no one is clearing MGMT_CLUSTER_NAME when using KIND, so this is always going to be true. Improve this logic.
```
[the remark this answers is not in the document]
Comment (nawazkh) [reply Tiltfile:412]: Yes, good call out, will update the epic with it.

**Comment by nawazkh** [comment]
> Should we have somewhere to document which CIDR ranges are "reserved" by which flavors? If we add a new flavor, we'd want to make sure it doesn't overlap, and it would be better not to have that rely just on maintainers' memories.

Where could be a good place to put this?
Does a readme in /templates/flavors be a good place to put the CIDR matrix ?

**Comment by nawazkh** [comment]
Screenshot 2024-11-18 at 3 33 46 PM
Something is up with GitHub, it is not letting me see my changes where I dropped my commits to just 5 instead of 23.

File: Tiltfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (nawazkh) [reply Tiltfile:412]: Created https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/5289

**Comment by nawazkh** [comment]
> > Should we have somewhere to document which CIDR ranges are "reserved" by which flavors? If we add a new flavor, we'd want to make sure it doesn't overlap, and it would be better not to have that rely just on maintainers' memories.
> 
> Where could be a good place to put this? Does a readme in /templates/flavors be a good place to put the CIDR matrix ?

Updated https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/5288/files#diff-e7ce6ca4dcc711afa1a2e1afdb50d9c9902749a29f4af50cde1a12f2e51f4858 with the CIDR matrix

**Pull request closed without merging**